pkg/frontend/querymiddleware/running.go

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+	"github.com/opentracing/opentracing-go/ext"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+func NewFrontendRunningRoundTripper(next http.RoundTripper, service services.Service, timeout time.Duration, log log.Logger) http.RoundTripper {
+	return &frontendRunningRoundTripper{
+		next:    next,
+		service: service,
+		timeout: timeout,
+		log:     log,
+	}
+}
+
+type frontendRunningRoundTripper struct {
+	next    http.RoundTripper
+	service services.Service
+	timeout time.Duration
+	log     log.Logger
+}
+
+func (f *frontendRunningRoundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
+	if err := awaitQueryFrontendServiceRunning(request.Context(), f.service, f.timeout, f.log); err != nil {
+		return nil, apierror.New(apierror.TypeUnavailable, err.Error())
+	}
+
+	return f.next.RoundTrip(request)
+}
+
+// This method is not on frontendRunningRoundTripper to make it easier to test this logic.
+func awaitQueryFrontendServiceRunning(ctx context.Context, service services.Service, timeout time.Duration, log log.Logger) error {
```

Comment:
(nit) couldn't this be generalised as it can be applied to any services.Service, not just query frontend?

Reply:
It could, but at the moment we don't have a need for it to be used elsewhere.